Konami (GX) blending - additive sprite blending, improved tile blending

FredYeye opened the pull request:
I feel like I'm starting to amass a lot of changes so I'm opening a PR now. That, and the current changes are starting to reach a stable state! ...I hope!

The TL;DR version, games I know to be affected in parentheses:
**Additive sprite blending** (mmaulers)
**Improved tile blending** (mystwarr, viostorm, metamrph, sexyparo)

---

10 months ago, master Galibert sent down His [holy](https://github.com/orgs/mamedev/discussions/109#discussioncomment-9355919) [scriptures](https://github.com/orgs/mamedev/discussions/109#discussioncomment-9429978) to a certain inquisitive peasant. Since then I've spent some time on and off trying to interpret His writings, and well, I can confidently say that *some* of it has definitely sunk in!

---

File by file commentary:

**k053246_k053247_k055673.cpp, k053246_k053247_k055673.h:**
I went back to update zdrawgfxzoom32GP(), now down to a third of its original size before I started changing it. As a thinly veiled excuse for reworking the whole function, I went ahead and added additive sprite blending (to the now singular alpha call site).

Remarks: the mix priority setting is not yet handled. It seems simple enough (flip dst and src), but I would like to find an example of this before I implement it.

**k054156_k054157_k056832.cpp:**
The `attr` variable holds what appear to be the elusive tile (external) mix codes. Attach it to the `flags` variable so these bits can be accessed in the tile callback functions. Tiles with mix codes gets their own tilemap category.

Remarks: ~~this is kind of a hacky way to get `attr` into the callbacks. This should at minimum be documented, or be changed.~~ I've now changed the callback to include an attr param.

**k054338.cpp:**
Update set_alpha_level. This function now returns a level, an additive blend bool and a mixpri bool. Minor style changes to the overall file.

Remarks: set_alpha_level doesn't actually *set* anything. Maybe rename to get_alpha_level?

**moo.cpp, xexex.cpp:**
Mask out the new additive & mixpri bits from set_alpha_level calls for now, until it's known if / how they should be used over there.

**mystwarr_v.cpp, mystwarr.h:**
Remove mystwarr water hack.
Update mystwarr_tile_callback (and add viostorm_tile_callback) to read tile mix codes, store last read mix code in a new m_last_alpha_tile_mix_code variable.
Attach m_last_alpha_tile_mix_code to `mixerflags`, which happens to have two unused bits.

Remarks: ~~I should document that the two last bits of `mixerflags` is carrying a tile mix code now.~~ I updated the mixerflags documentation to mention the usage of the last two bits.

**konamigx_v.cpp, konamigx.cpp, konamigx.h:**
Shrink GX_MAX_SPRITES, which to the best of my ability seems to be oversized. There does seem to be several oversized arrays / defs in these files, so I think this is one of them.
The usual *FredYeye* updates to konamigx_mixer - move declarations closer to use, more suited types, rename `temp` vars, etc.
Change objpool to a vector, simplifying usage (push_back, size).
Replace sorting loop with reverse + stable_sort.
Improve gx_draw_basic_tilemaps - read internal / external alpha mix codes based on vinmix_on. Tiles with mixcodes get drawn in a separate pass for per-tile blending.
Update alpha_tile_callback and add salmndr2_tile_callback, same as the mystwarr_v callbacks.

Remarks: ~~There are still some things that should be done here - alpha_tile_callback needs to be cleaned up, for example.~~ Mostly done here now.

---

What started out as trying to sort out konamigx_mixer() to look at shadow/priority issues instead ended up with me finally getting a foot into the figurative tile blending door. The mystwarr water hack is gone, and sexyparo gets transparent windows. This might affect many GX and related games. Alpha blending might be broken in some games now, and needs to get their mix codes attached in their respective callbacks. salmndr2 got tagged in my automatic video comparison for differing from earlier versions, that's why I managed to fix it already.

Known problems:
metamrph: stained glass windows are near-transparent at the moment. I think additive tile blending will fix it...
viostorm: character names fade in in reverse. This also uses additive blending, so this might also get fixed once that's in.
fantjour: the top & bottom flames at the captain kebab ship go missing. Uses additive blending.

---

I'll mark this as a draft and go ahead and try to improve on some of the current things that are left. Thanks for coming to my ted talk.

galibert commented:
Please ping me at galibert@pobox.com with an email I can reach you at.

FredYeye commented:
@cuavas i hope you don't mind the ping. I'd like to attempt adding additive blending for tiles. I guess what I would like to know is where I should be doing this?
Tiles are currently being drawn in `k054156_k054157_k056832.cpp`, function `m_tilemap_draw()`, by using `tmap->draw(screen, bitmap, drawrect, flags, priority);`. So drawing is currently being done inside `tilemap.cpp`. I don't know if I'm supposed to add additive blending functionality to this file, or somewhere else (`k054156_k054157_k056832.cpp`?). I figured it's maybe best to ask about these kinds of things...

TheCoolPup commented:
did some testing and most of the graphical bugs are gone now in mystic warriors, only bugs left is this bug and a brightness issue when going to the skull tank 
![Mystic Warriors (ver JAA)  mystwarrj  - MAME 0 275 (LLP64) 3_17_2025 10_01_19 PM](https://github.com/user-attachments/assets/4dfe4617-11bf-49bf-9ec9-d10f141b3073)
![gif-output-Tue18Mar2025021036GMT](https://github.com/user-attachments/assets/65e2b000-507a-4e2b-9646-5fa4c104c03b)
the screen is supposed to be getting darker the closer you get

FredYeye commented:
Oh, yeah, you're right! It does get darker, but it doesn't seem to get dark enough? I'll have to look at it

FredYeye commented:
I think I'm done adding new stuff to this PR now. I've opted to add a hack for additive alpha (invert alpha value) so it looks fine until it's properly handled.

While the sexyparo castle windows are transparent now, the transparency seemingly turns off for a frame or two when picking up power ups. I need to check PCB footage to see if that's actually a thing or not, but most videos are high score videos so they don't go this route.

FredYeye commented:
> There's also a severe graphical regression in Martial Champion (mtlchamp).

Thanks for pointing that out! I completely overlooked that one. This should preferably be fixed before a potential merge.

> I'm not sure if the need to fix those is within the scope of this pull request, but are there any considerations regarding that?

Yes, I'm at least planning to keep addressing various issues with GX (and related) games.

> And as for the stage transition effects in Violent Storm (viostorm), which were slightly broken in 0.275, do those now render correctly so [MT9131](https://mametesters.org/view.php?id=9131) can be marked fixed once this pull request is merged?

I think this is still broken. I was planning on tackling this, but managed to get tile blending working instead, heh. I will definitely revisit this issue (shadow priority problem?).

TheCoolPup commented:
Bro is about to become the golden savior of konami games on mame haha
![Uploading GIpr4AvW8AAUIrU.jpg…]()

TheCoolPup commented:
tested the new build and still doesnt get as dark as it should, but still i cant be too upset considering almost everything else has been fixed

TheCoolPup commented:
yet another regression 
![Mystic Warriors (ver JAA)  mystwarrj  - MAME 0 275 (LLP64) 3_18_2025 9_54_03 PM](https://github.com/user-attachments/assets/3a52b743-9b08-499f-a802-d30c82be6eff)

TheCoolPup commented:
noticed this too 
![Monster Maulers (ver EAA)  mmaulers  - MAME 0 275 (LLP64) 3_18_2025 9_58_15 PM](https://github.com/user-attachments/assets/c20a60d4-ad1c-4482-a783-de9fd56a221d)

JimCarlTay commented:
So, were those screenshots referring to certain stages in Mystic Warriors (mystwarr) and Monster Maulers (mmaulers) being too dark?

I was more concerned about the graphics regression in Martial Champion (mtlchamp).

FredYeye commented:
> yet another regression ![Mystic Warriors (ver JAA) mystwarrj - MAME 0 275 (LLP64) 3_18_2025 9_54_03 PM](https://private-user-images.githubusercontent.com/91135037/424259794-3a52b743-9b08-499f-a802-d30c82be6eff.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIzNzM4NTYsIm5iZiI6MTc0MjM3MzU1NiwicGF0aCI6Ii85MTEzNTAzNy80MjQyNTk3OTQtM2E1MmI3NDMtOWIwOC00OTlmLWE4MDItZDMwYzgyYmU2ZWZmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE5VDA4MzkxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFlYzA4NDgxZjc2MDQ4MzkwMDA5NTEwOGU5Y2I5MDcxMzIwY2M3YWFiNDZhN2FjNGE3NmEwNmVhOTZhMTA1NGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.dAwVErlMj5jYXCS9Om4z8zPm3OIYpzIp_avF-OMsLkQ)

Hmm, I can't repro this one. This part looks fine with the latest changes for me?

TheCoolPup commented:
And with monster maulers you can notice those lines in his power up graphic thing

TheCoolPup commented:
ok so the background image was due to the cheat 
![Mystic Warriors (ver JAA)  mystwarrj  - MAME 0 275 (LLP64) 3_19_2025 2_13_56 PM](https://github.com/user-attachments/assets/2934d1a1-1fd6-4578-80d1-f1594323b3e5)

TheCoolPup commented:
getting here normally does not cause this issue

TheCoolPup commented:
https://youtu.be/hBtjrd3qYdw also got some hd footage of the pcb for you

TheCoolPup commented:
Just posted a week ago oddly enough

TheCoolPup commented:
did something happen with the fixes? they dont seem to be in 0.276

JimCarlTay commented:
It's because this pull request isn't merged before March 22.

happppp commented:
Some small things with source code consistency:
- braces style: in mystwarr_v.cpp you added some new code with K&R braces but the rest of the file is Allman style
- shorthand variable types: if the rest of the file has uint8_t, use that, not u8
- lower/uppercase hex: unless the rest of the file has uppercase hex (which is not likely in MAME), use lowercase, so change 0xFF that you added to 0xff

FredYeye commented:
> braces style: in mystwarr_v.cpp you added some new code with K&R braces but the rest of the file is Allman style

I think I've done this mistake in every mame PR I've ever made...

> shorthand variable types: if the rest of the file has uint8_t, use that, not u8

I've changed `uX` to `uintX_t` in some files where it was obvious to do so. Some of these files feature mixed usage already, so I didn't change anything there. Let me know if I should change those to.

> lower/uppercase hex: unless the rest of the file has uppercase hex (which is not likely in MAME), use lowercase, so change 0xFF that you added to 0xff

Fixed.

rb6502 commented:
@FredYeye Just as a reference, on the uint8_t vs u8 thing, we accept both, but they should all be consistent within the same file.  I've been generally converting things to the shorter types when touching them, but that's my preference and not a rule.
